### PR TITLE
LUCENE-9615: Expose HnswGraphBuilder index-time hyperparameters as FieldType attributes

### PR DESCRIPTION
**Pull request by sbeniwal12**
Expose HnswGraphBuilder index-time hyperparameters as FieldType attributes

# Description

HnswGraphBuilder has a few tunables: maxConnections, beamWidth, and we may add a few more, such as whether to use a diversity heuristic when choosing neighbors to link in the graph. Currently these are locked to defaults set by global variables. Instead we should provide some interface for configuring them.

# Solution

Here with this change we allow to pass down these parameters as FieldType attributes of VectorFields as well provide a convenience method in VectorFields.java to  add these parameters in FieldType attributes. The value passed in the attributes would be used in `Lucene90VectorWriter ` while creating a HNSW graph, in case no value is found we switch back to default parameter values present in `HnswGraphBuilder`. 

# Tests

Updated way of setting these parameters  in `KnnGraphTester.java` for testing indexing and search performance of a knn-graph and added unit tests for new constructor and public method in `VectorField.java` in `TestVectorValues.java`.

# Checklist

Please review the following and check all that apply:

- [x] I have reviewed the guidelines for [How to Contribute](https://wiki.apache.org/solr/HowToContribute) and my code conforms to the standards described there to the best of my ability.
- [x] I have created a Jira issue and added the issue ID to my pull request title.
- [x] I have given Solr maintainers [access](https://help.github.com/en/articles/allowing-changes-to-a-pull-request-branch-created-from-a-fork) to contribute to my PR branch. (optional but recommended)
- [x] I have developed this patch against the `master` branch.
- [x] I have run `./gradlew check`.
- [x] I have added tests for my changes.
- [ ] I have added documentation for the [Ref Guide](https://github.com/apache/lucene-solr/tree/master/solr/solr-ref-guide) (for Solr changes only).

### CODE REVIEW

File: lucene/core/src/java/org/apache/lucene/codecs/lucene90/Lucene90VectorWriter.java [hunk 1]
```diff
@@ -188,9 +190,29 @@ private void writeGraph(
...
-      int count)
+      int count,
+      String maxConnStr,
+      String beamWidthStr)
       throws IOException {
-    HnswGraphBuilder hnswGraphBuilder = new HnswGraphBuilder(vectorValues);
+    int maxConn, beamWidth;
+    if (maxConnStr == null) {
+      maxConn = HnswGraphBuilder.DEFAULT_MAX_CONN;
+    } else if (!maxConnStr.matches("[0-9]+")) {
```
Comment (msokolov) [inline lucene/core/src/java/org/apache/lucene/codecs/lucene90/Lucene90VectorWriter.java:200]: I don't think we need this - we can allow parseInt to throw an exception. Let's catch `NumberFormatException` and rethrow with more context (which attribute caused the exception). Also HnswGraphBuilder tests for `<= 0`, so we don't need to check that here.

File: lucene/core/src/java/org/apache/lucene/document/VectorField.java [hunk 2]
```diff
@@ -53,6 +54,44 @@ private static FieldType getType(float[] v, VectorValues.SearchStrategy searchSt
...
+      int dimension, VectorValues.SearchStrategy searchStrategy, int maxConn, int beamWidth) {
+    if (dimension == 0) {
+      throw new IllegalArgumentException("cannot index an empty vector");
+    }
+    if (dimension > VectorValues.MAX_DIMENSIONS) {
+      throw new IllegalArgumentException(
+          "cannot index vectors with dimension greater than " + VectorValues.MAX_DIMENSIONS);
+    }
+    if (searchStrategy == null) {
+      throw new IllegalArgumentException("search strategy must not be null");
```
Comment (msokolov) [inline lucene/core/src/java/org/apache/lucene/document/VectorField.java:77]: Let's also assert `searchStrategy.isHnsw()` to catch attempts to use `NONE` or some other unsupported future strategy.

File: lucene/core/src/test/org/apache/lucene/util/hnsw/KnnGraphTester.java [hunk 3]
```diff
@@ -132,13 +135,13 @@ private void run(String... args) throws Exception {
           if (iarg == args.length - 1) {
             throw new IllegalArgumentException("-beamWidthIndex requires a following number");
           }
-          HnswGraphBuilder.DEFAULT_BEAM_WIDTH = Integer.parseInt(args[++iarg]);
```
Comment (msokolov) [inline lucene/core/src/test/org/apache/lucene/util/hnsw/KnnGraphTester.java:135]: With this change, we no longer have any need to make these static variables writable - let's change them to `final` in `HnswGraphBuilder`

File: lucene/core/src/java/org/apache/lucene/codecs/lucene90/Lucene90VectorWriter.java (hunk 1 above)
Comment (sbeniwal12) [reply lucene/core/src/java/org/apache/lucene/codecs/lucene90/Lucene90VectorWriter.java]: Thanks for pointing out, threw `NumberFormatException` with message describing which attribute caused the exception.

File: lucene/core/src/java/org/apache/lucene/document/VectorField.java (hunk 2 above)
Comment (sbeniwal12) [reply lucene/core/src/java/org/apache/lucene/document/VectorField.java]: Added this check and also added a unit test for this check.

File: lucene/core/src/test/org/apache/lucene/util/hnsw/KnnGraphTester.java (hunk 3 above)
Comment (sbeniwal12) [reply lucene/core/src/test/org/apache/lucene/util/hnsw/KnnGraphTester.java:135]: Made them final and also made changes to `TestKnnGraph.java` to accommodate this change.

**Review by msokolov — APPROVED** [review]
Looks good, thanks @sbeniwal12 !

**Pull request merged**